README.md
Description
===========
Installs Twitter's Storm distributed computation platform.  Includes recipes for installing
both the Nimbus / Web UI component and the Supervisor component.

Requirements
============

* java cookbook
* runit cookbook

Attributes
==========

Usage
=====

This recipe relies on two setup components that need to be noted as they are not used
in many (or any) community cookbooks.



CHANGELOG.md
## 1.2.27
* Add search capability

## 1.2.6
* Cleaned up food critic warnings
* Made default version 0.8.2

## 1.2.5
* Fixing path in supervisor runit script

## 1.2.4
* Added conf creation to the list of dirs to creat

## 1.2.3
* Changed default ZooKeeper Settings

## 1.2.2
* Removed deploy_build block entirely 
* Changed remote_file to use :create_if_missing

## 1.2.1
* Updated run_list to recipe for determing nimbus node

## 1.2.0
* Changes to set node data for topologies to use

## 1.1.0
* Added runit include

## 1.0.34
* Used joshes correct changes

## 1.0.33
* Changed template to be current
* Changed the service reload to not be immediate

## 1.0.31
* Added support for new runit version.

## 1.0.30
* Changed supervisor control script which would prevent a supervisor from
  starting because worker processes were already running. It can happen
  that a supervisor will die but the workers not, then runit will no
  longer be able to start the supervisor. This fix should prevent that.

## 1.0.29
* Changed the default worker memory size.

## 1.0.28
* Add support for Ubuntu to the metadata
* Add a Readme
* Increase file handler limit for nimbus to 20,000

## 1.0.27
* Bump the file handler limit on the webui to 1024. The limit was already increased on Nimbus

## 1.0.26
* Cleans out the state of only the previous runs supervisors, but not the workers.
  Removing the workers state caused supsequent supervisors to bitch about not
  being able to remove the pid files from /mnt/storm/work/XXX/pids without
  end.

## 1.0.25
* Remove the force-yes for installing packages.  This was needed due to the old setup of our apt repo
* Adding default attribute for the storm version

## 1.0.24
* Removed the 'force-stop' call since it is not suported by the 'service' command

## 1.0.23
* Added undeploy scripts.

## 1.0.22
* Address food critic warnings
* Remove the ulimits template since this doesn't work in Ubuntu and was replaced with a ulimit statement in 1.0.21
* Don't fall back to attributes if the Zookeeper node search fails

## 1.0.21
* Add ulimit statement to the runit script for nimbus to increase the file limit to 10240

## 1.0.20
* Add log_dir and install_dir as default attributes

## 1.0.19
* Changed the limits.d file to bump the open file limit for any user

## 1.0.18
* Removed  specific attributes

## 1.0.17
* Change nimbus and storm ui start/stop scripts to kill related
  processes

## 1.0.16
* Increase file limits for the storm user from 1024 to 32k

## 1.0.15
* Someone changes something

## 1.0.14
* Create a link /opt/storm/current that points to the current version


recipes/zookeeper.rb
package 'java-1.7.0-openjdk-devel'
package 'bison'
package 'flex'
package 'gcc'
package 'gcc-c++'
package 'kernel-devel'
package 'make'
package 'm4'
package 'patch'

remote_file "#{Chef::Config[:file_cache_path]}/zookeeper-#{node[:zookeeper][:version]}.tar.gz" do
  source "#{node[:zookeeper][:download_url]}/zookeeper-#{node[:zookeeper][:version]}/zookeeper-#{node[:zookeeper][:version]}.tar.gz"
  mode   0655
end

directory "#{node[:zookeeper][:root_dir]}" do
  user "#{node[:zookeeper][:user]}"
  mode 00755
  recursive true
end

directory "#{node[:zookeeper][:data_dir]}" do
  user "#{node[:zookeeper][:user]}"
  mode 00755
  recursive true
end

execute "install_zookeeper" do
  cwd "#{node[:zookeeper][:root_dir]}"
  command "tar -zxvf #{Chef::Config[:file_cache_path]}/zookeeper-#{node[:zookeeper][:version]}.tar.gz"
  not_if { ::File.exists?("#{node[:zookeeper][:root_dir]}/zookeeper-#{node[:zookeeper][:version]}") }
end

template "setup_zookeeper_config" do
  path "#{node[:zookeeper][:root_dir]}/zookeeper-#{node[:zookeeper][:version]}/conf/zoo.cfg"
  source "zoo.cfg.erb"
  owner node[:zookeeper][:user]
  group node[:zookeeper][:group]
  mode 0744
  variables(
    :data_dir => node['zookeeper']['data_dir'],
    :clientPort => node['zookeeper']['clientPort']
  )
  action :create_if_missing
end

bash "start_zookeeper" do
  user node[:zookeeper][:user]
  cwd "#{node[:zookeeper][:data_dir]}"
  code <<-EOH
    #{node[:zookeeper][:root_dir]}/zookeeper-#{node[:zookeeper][:version]}/bin/zkServer.sh start >>zookeeper.log 2>&1
  EOH
end


recipes/dependencies.rb

include_recipe "java::oracle"

node['storm']['packages'].each do |pkg|
  package pkg
end

bash "flush iptables" do
  command "service iptables stop"
end

if node['zookeeper']['required']
  include_recipe "storm::zookeeper"
end

if node['zeromq']['required']
  remote_file "#{Chef::Config[:file_cache_path]}/zeromq-#{node[:zeromq][:version]}.tar.gz" do
    source "#{node['zeromq']['download_url']}/zeromq-#{node[:zeromq][:version]}.tar.gz"
    mode   0744
    action :create_if_missing
  end

  execute "tar" do
    cwd     "#{Chef::Config[:file_cache_path]}"
    command "tar -zxvf #{Chef::Config[:file_cache_path]}/zeromq-#{node[:zeromq][:version]}.tar.gz"
  end

  bash "build_zeromq" do
    cwd "#{Chef::Config[:file_cache_path]}/zeromq-#{node[:zeromq][:version]}"
    code <<-EOH
      (./configure && make)
    EOH
  end

  bash "install_zeromq" do
    cwd "#{Chef::Config[:file_cache_path]}/zeromq-#{node[:zeromq][:version]}"
    code <<-EOH
      make install
      ldconfig
    EOH
  end
end

if node['jzmq']['required']
 
  remote_file "#{Chef::Config[:file_cache_path]}/jzmq-master.zip" do
    source "#{node['jzmq']['download_url']}"
    action :create_if_missing
  end

  execute "unzip" do
    cwd     "#{Chef::Config[:file_cache_path]}"
    command "unzip #{Chef::Config[:file_cache_path]}/jzmq-master.zip"
  end

  bash "build_jzmq" do
    cwd "#{Chef::Config[:file_cache_path]}/jzmq-master"
    code <<-EOH
      (./autogen.sh && ./configure)
    EOH
  end

  bash "install_jzmq" do
    cwd "#{Chef::Config[:file_cache_path]}/jzmq-master"
    code <<-EOH
      make install
    EOH
  end
end


recipes/default.rb
include_recipe "storm::dependencies"
include_recipe "partial_search"

# Get list of zookeeper hosts
node.override['storm']['zookeeper']['hosts'] = search(:node,
                                                      "chef_environment:#{node.chef_environment} AND role:#{node['storm']['storm_zookeeper_role']}",
                                                      :filter_result => {'ipaddress' => ['ipaddress']})

node.override['storm']['nimbus']['host'] = search(:node,
                                                  "chef_environment:#{node.chef_environment} AND role:#{node['storm']['storm_nimbus_role']}",
                                                  :filter_result => {'ipaddress' => ['ipaddress']}).first


if node['storm']['zookeeper']['hosts'].nil?
  Chef::Application.fatal!('No Zookeeper nodes found')
end

if node['storm']['nimbus']['host'].nil?
  Chef::Application.fatal!('No Nimbus node found')
end

storm_remote_name = "#{node['storm']['download_url']}#{node['storm']['download_dir']}"

remote_file "#{Chef::Config[:file_cache_path]}/apache-storm-#{node[:storm][:version]}.tar.gz" do
  # source "#{node['storm']['download_url']}/apache-storm-#{node[:storm][:version]}/apache-storm-#{node[:storm][:version]}.tar.gz"
  source storm_remote_name
  action :create_if_missing
end

directory node['storm']['root_dir'] do
  user "#{node[:storm][:user]}"
  mode 00755
  recursive true
  action :create
  not_if { ::File.directory?("#{node[:storm][:root_dir]}") }
end

execute "extract" do
  cwd     node['storm']['root_dir']
  command "tar -xzvf #{Chef::Config[:file_cache_path]}/apache-storm-#{node[:storm][:version]}.tar.gz"
  not_if { ::File.directory?("#{node[:storm][:root_dir]}/apache-storm-#{node[:storm][:version]}") }
end

directory node['storm']['data_dir'] do
  user "#{node[:storm][:user]}"
  mode 00755
  recursive true
  action :create
  not_if { ::File.directory?("#{node[:storm][:data_dir]}") }
end

template "setup_storm_config" do
  path "#{node[:storm][:root_dir]}/apache-storm-#{node[:storm][:version]}/conf/storm.yaml"
  source "storm.yaml.erb"
  owner node[:storm][:user]
  group node[:storm][:group]
  mode 0744
end

template "/tmp/setup_self_in_hosts.sh" do
  source "conf_self_in_hosts.sh"
  mode 0755
end

execute "setup hosts file" do
  command "sh /tmp/setup_self_in_hosts.sh"
end
